packages/belos/tpetra/src/solvers/Belos_Tpetra_GmresSingleReduce.hpp

```diff
@@ -324,7 +723,7 @@ class GmresSingleReduce : public Gmres<SC, MV, OP> {
       else if (output.numIters < input.maxNumIters) {
         // Restart, only if max inner-iteration was reached.
         // Otherwise continue the inner-iteration.
-        if (iter >= restart) {
+        if (iter >= restart || H(iter,iter-1) == zero) { // done with restart cycyle, or probably lost ortho
```

Comment:
Should there be a tolerance here?  How likely is it that H(i,i-1) = 0 exactly?  Also, should it be H(iter, iter+1)?? 
If we restart early, it might also be nice to give a warning to the user.

Reply:
I believe we check the tolerance inside the orthogonalization kernel, and set H(i,i-1) to be zero if failed (e.g., [here](https://github.com/trilinos/Trilinos/blob/c27e31e8397f5bdf4d7861be118f01e23f522faf/packages/belos/tpetra/src/solvers/Belos_Tpetra_GmresSingleReduce.hpp#L426) But, we probably should take the tolerance as an input parameter..

Reply:
Okay.  That sounds reasonable.  I still think we should warn the user (print a message) if orthog fails and we restart early.

Reply:
Thank you, @jennloe . I added a couple of print warnings for the non-positive scaling factors!!

Reply:
Looks good!